core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -393,18 +395,20 @@ class FullFetchContext(private val time: Time,
   * @param reqMetadata  The request metadata.
   * @param session      The incremental fetch request session.
   */
-class IncrementalFetchContext(private val time: Time,
-                              private val reqMetadata: JFetchMetadata,
-                              private val session: FetchSession) extends FetchContext {
+final class IncrementalFetchContext(private val time: Time,
+                                    private val reqMetadata: JFetchMetadata,
+                                    private val session: FetchSession) extends FetchContext {
 
   override def getFetchOffset(tp: TopicPartition): Option[Long] = session.getFetchOffset(tp)
 
-  override def foreachPartition(fun: (TopicPartition, FetchRequest.PartitionData) => Unit): Unit = {
+  override def partitions: mutable.Map[TopicPartition, FetchRequest.PartitionData] = {
     // Take the session lock and iterate over all the cached partitions.
     session.synchronized {
-      session.partitionMap.iterator.asScala.foreach(part => {
-        fun(new TopicPartition(part.topic, part.partition), part.reqData)
-      })
+      val copy = mutable.Map.newBuilder[TopicPartition, FetchRequest.PartitionData]
+      session.partitionMap.iterator.asScala.foreach { part =>
+        copy += (new TopicPartition(part.topic, part.partition) -> part.reqData)
+      }
+      copy.result()
```

Comment:
@cmccabe  I opted to perform a copy here to avoid executing unknown code while holding a lock.

Reply:
Move all of the code to `KafkaApis`. This PR doesn't introduce any new files.

Reply:
Please do not copy all the partition data structures.  There may be a lot of them.